test/governon.flow.test.ts

```diff
@@ -0,0 +1,374 @@
+      const proposal1Description = "Mint additional tokens to voter1";
+
+      const proposal2Targets = [tokenAddr];
+      const proposal2Values = [0];
+      const proposal2Calldatas = [
+        token.interface.encodeFunctionData("mint", [voter2.address, proposal2Amount]),
+      ];
+      const proposal2Description = "Mint additional tokens to voter2";
+
+      await governance.connect(proposer).propose(
+        proposal1Targets,
+        proposal1Values,
+        proposal1Calldatas,
+        proposal1Description
+      );
+      await governance.connect(proposer).propose(
+        proposal2Targets,
+        proposal2Values,
+        proposal2Calldatas,
+        proposal2Description
+      );
+
+      const proposal1DescriptionHash = ethers.keccak256(ethers.toUtf8Bytes(proposal1Description));
+      const proposal2DescriptionHash = ethers.keccak256(ethers.toUtf8Bytes(proposal2Description));
+      const proposal1Id = await governance.hashProposal(
+        proposal1Targets,
+        proposal1Values,
+        proposal1Calldatas,
+        proposal1DescriptionHash
+      );
+      const proposal2Id = await governance.hashProposal(
+        proposal2Targets,
+        proposal2Values,
+        proposal2Calldatas,
+        proposal2DescriptionHash
+      );
+
+      await mineBlocks(2);
+
+      await governance.connect(proposer).castVote(proposal1Id, 1);
+      await governance.connect(voter1).castVote(proposal1Id, 1);
+      await governance.connect(voter2).castVote(proposal1Id, 0);
+
+      await governance.connect(proposer).castVote(proposal2Id, 1);
+      await governance.connect(voter1).castVote(proposal2Id, 0);
+      await governance.connect(voter2).castVote(proposal2Id, 1);
+
+      await mineBlocks(5);
+
+      expect(await governance.state(proposal1Id)).to.equal(4);
+      expect(await governance.state(proposal2Id)).to.equal(4);
+
+      await governance.connect(proposer).queue(
+        proposal1Targets,
+        proposal1Values,
+        proposal1Calldatas,
+        proposal1DescriptionHash
+      );
+      await governance.connect(proposer).queue(
+        proposal2Targets,
+        proposal2Values,
+        proposal2Calldatas,
+        proposal2DescriptionHash
+      );
+
+      await ethers.provider.send("evm_increaseTime", [3600]);
+      await ethers.provider.send("evm_mine", []);
+
+      await governance.connect(proposer).execute(
+        proposal1Targets,
+        proposal1Values,
+        proposal1Calldatas,
+        proposal1DescriptionHash
+      );
+      await governance.connect(proposer).execute(
+        proposal2Targets,
+        proposal2Values,
+        proposal2Calldatas,
+        proposal2DescriptionHash
+      );
+
+      // = start balance - `proposal${n}Amount`
+      expect(
+        await token.balanceOf(voter1.address)
+      ).to.equal(
+        ethers.parseUnits("550")
+      );
+
+      expect(
+        await token.balanceOf(voter2.address)
+      ).to.equal(
+        ethers.parseUnits("575")
+      );
```

Comment:
how are voting tokens used? what happens with them?
can you vote with 0 token? can you vote if not delegated any tokens?
can you vote if have tokens but none delegated?